File: Tween/Connection/TwitVideo.vb
Imports System.IO
Imports System.Text
Imports System.Net

Public Class TwitVideo
    Inherits HttpConnection

    Private Const PostMethod As String = "POST"
    Private Const GetMethod As String = "GET"
    Private Const ConsumerKey As String = "c72b8327466ad675782538f7c869738d"
    Private multimediaExt() As String = {".avi", _
                                         ".wmv", _
                                         ".flv", _
                                         ".m4v", _
                                         ".mov", _
                                         ".mp4", _
                                         ".rm", _
                                         ".mpeg", _
                                         ".mpg", _
                                         ".3gp", _
                                         ".3g2"}
    Private pictureExt() As String = {".jpg", _
                                    ".jpeg", _
                                    ".gif", _
                                    ".png"}

    Public Function Upload(ByVal mediaFile As FileInfo, _
                           ByVal message As String, _
                           ByVal keyword As String, _
                           ByVal username As String, _
                           ByVal twitter_id As String, _
                           ByRef content As String) As HttpStatusCode
        'Message必須
        If String.IsNullOrEmpty(message) Then Throw New ArgumentException("'Message' is required.")
        'Check filetype and size
        If Array.IndexOf(multimediaExt, mediaFile.Extension.ToLower) > -1 Then
            If mediaFile.Length > 20971520 Then Throw New ArgumentException("File is too large.")
        ElseIf Array.IndexOf(pictureExt, mediaFile.Extension.ToLower) > -1 Then
            If mediaFile.Length > 10485760 Then Throw New ArgumentException("File is too large.")
        Else
            Throw New ArgumentException("Service don't support this filetype.")
        End If
        'Endpoint(URI+Token)
        Const URLBASE As String = "http://api.twitvideo.jp/oauth/upload/"
        Dim data As Byte() = Encoding.ASCII.GetBytes(ConsumerKey.Substring(0, 9) + username)
        Dim bHash As Byte() = (New System.Security.Cryptography.MD5CryptoServiceProvider()).ComputeHash(data)
        Dim url As String = URLBASE + BitConverter.ToString(bHash).ToLower.Replace("-", "")
        'Parameters
        Dim param As New Dictionary(Of String, String)
        param.Add("username", username)
        If Not String.IsNullOrEmpty(twitter_id) Then param.Add("twitter_id", twitter_id)
        If Not String.IsNullOrEmpty(keyword) Then param.Add("keyword", keyword)
        param.Add("type", "xml")
        param.Add("message", message)
        Dim binary As New List(Of KeyValuePair(Of String, FileInfo))
        binary.Add(New KeyValuePair(Of String, FileInfo)("media", mediaFile))
        Me.InstanceTimeout = 60000     'タイムアウト60秒

        Dim req As HttpWebRequest = CreateRequest(PostMethod, _
                                                  New Uri(url), _
                                                  param, _
                                                  binary, _
                                                  False)
        Return Me.GetResponse(req, content, Nothing, False)
    End Function

    Public Function CheckValidExtension(ByVal ext As String) As Boolean
        If Array.IndexOf(pictureExt, ext) > -1 Then
            Return True
        End If
        If Array.IndexOf(multimediaExt, ext) > -1 Then
            Return True
        End If
        Return False
    End Function

    Public Function GetFileType(ByVal ext As String) As UploadFileType
        If Array.IndexOf(pictureExt, ext) > -1 Then
            Return UploadFileType.Picture
        End If
        If Array.IndexOf(multimediaExt, ext) > -1 Then
            Return UploadFileType.MultiMedia
        End If
        Return UploadFileType.Invalid
    End Function

    Public Function GetFileOpenDialogFilter() As String
        Return "Image Files(*.gif;*.jpg;*.jpeg;*.png)|*.gif;*.jpg;*.jpeg;*.png|" + _
                "Movie Files(*.avi;*.wmv;*.flv;*.m4v;*.mov;*.mp4;*.rm;*.mpeg;*.mpg;*.3gp;*.3g2)|*.avi;*.wmv;*.flv;*.m4v;*.mov;*.mp4;*.rm;*.mpeg;*.mpg;*.3gp;*.3g2"
    End Function
End Class

File: OpenTween/Connection/TwitVideo.cs
// OpenTween - Client of Twitter
// Copyright (c) 2007-2011 kiri_feather (@kiri_feather) <kiri.feather@gmail.com>
//           (c) 2008-2011 Moz (@syo68k)
//           (c) 2008-2011 takeshik (@takeshik) <http://www.takeshik.org/>
//           (c) 2010-2011 anis774 (@anis774) <http://d.hatena.ne.jp/anis774/>
//           (c) 2010-2011 fantasticswallow (@f_swallow) <http://twitter.com/f_swallow>
//           (c) 2011      spinor (@tplantd) <http://d.hatena.ne.jp/spinor/>
// All rights reserved.
// 
// This file is part of OpenTween.
// 
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 3 of the License, or (at your option)
// any later version.
// 
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
// for more details. 
// 
// You should have received a copy of the GNU General Public License along
// with this program. If not, see <http://www.gnu.org/licenses/>, or write to
// the Free Software Foundation, Inc., 51 Franklin Street - Fifth Floor,
// Boston, MA 02110-1301, USA.

using HttpConnection = OpenTween.HttpConnection;
using HttpStatusCode = System.Net.HttpStatusCode;
using FileInfo = System.IO.FileInfo;
using ArgumentException = System.ArgumentException;
using Array = System.Array;
using Encoding = System.Text.Encoding;
using BitConverter = System.BitConverter;
using System.Collections.Generic; // for Dictionary<TKey, TValue>, List<T>, KeyValuePair<TKey, TValue>
using HttpWebRequest = System.Net.HttpWebRequest;
using Uri = System.Uri;
using UploadFileType = OpenTween.MyCommon.UploadFileType;
using MD5CryptoServiceProvider = System.Security.Cryptography.MD5CryptoServiceProvider;

namespace OpenTween
{
	public class TwitVideo : HttpConnection
	{
		private string[] multimediaExt = new string[] { ".avi", ".wmv", ".flv", ".m4v", ".mov", ".mp4", ".rm", ".mpeg", ".mpg", ".3gp", ".3g2" };

		private string[] pictureExt = new string[] { ".jpg", ".jpeg", ".gif", ".png" };

		private const long MaxPictureFileSize = 10 * 1024 * 1024;

		private const long MaxMultiMediaFileSize = 20 * 1024 * 1024;

		public HttpStatusCode Upload( FileInfo mediaFile, string message, string keyword, string username, string twitter_id, ref string content )
		{
			// Message必須
			if ( string.IsNullOrEmpty( message ) )
				throw new ArgumentException( "'Message' is required." );

			// Check filetype and size
			if ( Array.IndexOf( multimediaExt, mediaFile.Extension.ToLower() ) > -1 )
			{
                if ( mediaFile.Length > TwitVideo.MaxMultiMediaFileSize )
					throw new ArgumentException( "File is too large." );
			}
			else if ( Array.IndexOf( pictureExt, mediaFile.Extension.ToLower() ) > -1 )
			{
                if ( mediaFile.Length > TwitVideo.MaxPictureFileSize )
					throw new ArgumentException( "File is too large." );
			}
			else
			{
				throw new ArgumentException( "Service don't support this filetype." );
			}

			// Endpoint(URI+Token)
			const string URLBASE = "http://api.twitvideo.jp/oauth/upload/";
            byte[] data = Encoding.ASCII.GetBytes( ApplicationSettings.TwitVideoConsumerKey.Substring(0, 9) + username );

            byte[] bHash;
            using (MD5CryptoServiceProvider md5provider = new MD5CryptoServiceProvider())
            {
    	        bHash = md5provider.ComputeHash( data );
            }

			string url = URLBASE + BitConverter.ToString( bHash ).ToLower().Replace( "-", "" );
			// Parameters
			Dictionary< string, string > param = new Dictionary< string, string >();
			param.Add( "username", username );
			if ( !string.IsNullOrEmpty( twitter_id ) )
				param.Add( "twitter_id", twitter_id );
			if ( !string.IsNullOrEmpty( keyword ) )
				param.Add( "keyword", keyword );
			param.Add( "type", "xml" );
			param.Add( "message", message );
			List< KeyValuePair< string, FileInfo > > binary = new List< KeyValuePair< string, FileInfo > >();
			binary.Add( new KeyValuePair< string, FileInfo >( "media", mediaFile ) );
			this.InstanceTimeout = 60000; // タイムアウト60秒

			HttpWebRequest req = this.CreateRequest( HttpConnection.PostMethod, new Uri( url ), param, binary, false );
			return this.GetResponse( req, out content, null, false );
		}

		public bool CheckValidExtension( string ext )
		{
			if ( Array.IndexOf( this.pictureExt, ext.ToLower() ) > -1 )
				return true;

			if ( Array.IndexOf( this.multimediaExt, ext.ToLower() ) > -1 )
				return true;

			return false;
		}

		UploadFileType GetFileType( string ext )
		{
			if ( Array.IndexOf( this.pictureExt, ext.ToLower() ) > -1 )
				return UploadFileType.Picture;

			if ( Array.IndexOf( this.multimediaExt, ext.ToLower() ) > -1 )
				return UploadFileType.MultiMedia;

			return UploadFileType.Invalid;
		}

		bool IsSupportedFileType( UploadFileType type )
		{
			return type.Equals( UploadFileType.Picture ) || type.Equals( UploadFileType.MultiMedia );
		}

		long GetMaxFileSize( string ext )
		{
			if ( Array.IndexOf( this.pictureExt, ext.ToLower() ) > -1 )
                return TwitVideo.MaxPictureFileSize;

			if ( Array.IndexOf( this.multimediaExt, ext.ToLower() ) > -1 )
                return TwitVideo.MaxMultiMediaFileSize;

			return -1;
		}

		string GetFileOpenDialogFilter()
		{
			return "Image Files(*.gif;*.jpg;*.jpeg;*.png)|*.gif;*.jpg;*.jpeg;*.png|"
			+ "Movie Files(*.avi;*.wmv;*.flv;*.m4v;*.mov;*.mp4;*.rm;*.mpeg;*.mpg;*.3gp;*.3g2)|*.avi;*.wmv;*.flv;*.m4v;*.mov;*.mp4;*.rm;*.mpeg;*.mpg;*.3gp;*.3g2";
		}
	}
}
